ddtrace/internal/telemetry/writer.py

```diff
@@ -70,50 +88,68 @@ def send_event(self, request):
         headers["DD-Telemetry-Request-Type"] = request["request_type"]
+        headers["DD-Telemetry-API-Version"] = request["api_version"]
+        headers["DD-Agent-Hostname"] = request["host"]["hostname"]
+        if config.env:
+            headers["DD-Agent-Env"] = config.env
         return headers
 
 
-class TelemetryWriter(PeriodicService):
+class TelemetryBase(PeriodicService):
     """
-    Submits Instrumentation Telemetry events to the datadog agent.
-    Supports v1 of the instrumentation telemetry api
+    Common features of Telemetry services
     """
 
     # telemetry endpoint uses events platform v2 api
     ENDPOINT_V1 = "telemetry/proxy/api/v2/apmtelemetry"
 
-    def __init__(self):
-        # type: () -> None
-        super(TelemetryWriter, self).__init__(interval=_get_interval_or_default())
+    def __init__(self, interval):
+        # type: (float) -> None
+        super(TelemetryBase, self).__init__(interval=interval)
 
         # _enabled is None at startup, and is only set to true or false
         # after the config has been processed
         self._enabled = None  # type: Optional[bool]
+
         self._events_queue = []  # type: List[Dict]
-        self._integrations_queue = []  # type: List[Dict]
         self._lock = forksafe.Lock()  # type: forksafe.ResetObject
-        self._forked = False  # type: bool
+        forksafe.register(self._fork_writer)
+
         # Debug flag that enables payload debug mode.
         self._debug = asbool(os.environ.get("DD_TELEMETRY_DEBUG", "false"))
-        forksafe.register(self._fork_writer)
 
         # Counter representing the number of events sent by the writer. Here we are relying on the atomicity
         # of `itertools.count()` which is a CPython implementation detail. The sequence field in telemetry
         # payloads is only used in tests and is not required to process Telemetry events.
         self._sequence = itertools.count(1)
         self._client = _TelemetryClient(self.ENDPOINT_V1)
 
-    def _flush_integrations_queue(self):
-        # type: () -> List[Dict]
-        """Flushes and returns a list of all queued integrations"""
-        with self._lock:
-            integrations = self._integrations_queue
-            self._integrations_queue = []
-        return integrations
+    def add_event(self, payload, payload_type):
+        # type: (Dict[str, Any], str) -> None
+        """
+        Adds a Telemetry event to the TelemetryWriter event buffer
+
+        :param Dict payload: stores a formatted telemetry event
+        :param str payload_type: The payload_type denotes the type of telmetery request.
+            Payload types accepted by telemetry/proxy v1: app-started, app-closing, app-integrations-change
+        """
+        if self._enabled:
+            event = {
+                "tracer_time": int(time.time()),
+                "runtime_id": get_runtime_id(),
+                "api_version": "v1",
```

Comment:
Something to call out. I spoke to @mehulsonowal and it appears api v2 is not fully backwards compatible with v1. We should avoid bumping the telemetry writer to v2 in this PR (if possible) or scope api v2 to just metrics events.

Reply:
yes, I noticed if I update all telemetry requests to v2 system tests start to fail. I rollback to v1 and I tested manually and everything works fine. On the other hand, I split TelemetryWriter into TelemetryWriter (inverval of 60s and we use it to report heartbeat, startapp, close...) and TelemetryMetricsWriter (interval of 10s and to report generate-metrics and distributions). 

Thank you so much for your tip!